heron/common/src/cpp/config/topology-config-helper.cpp

```diff
@@ -399,17 +399,105 @@ bool TopologyConfigHelper::DropTuplesUponBackpressure(const proto::api::Topology
                                TopologyConfigVars::TOPOLOGY_DROPTUPLES_UPON_BACKPRESSURE, false);
 }
 
+// Return topology level config
+void TopologyConfigHelper::GetTopologyConfig(const proto::api::Topology& _topology,
+                                             std::map<sp_string, sp_string>& retval) {
+  if (_topology.has_topology_config()) {
+    const proto::api::Config& config = _topology.topology_config();
+    ConvertConfigToKVMap(config, retval);
+  }
+}
+
+// Update topology level config
+void TopologyConfigHelper::SetTopologyConfig(proto::api::Topology* _topology,
+                                             const std::map<sp_string, sp_string>& _update) {
+  if (_topology->has_topology_config()) {
+    proto::api::Config* config = _topology->mutable_topology_config();
+    UpdateConfigFromKVMap(config, _update);
+  }
+}
+
+// Return component level config
+void TopologyConfigHelper::GetComponentConfig(const proto::api::Topology& _topology,
+                                              const sp_string& _component_name,
+                                              std::map<sp_string, sp_string>& retval) {
+  // We are assuming component names are unique and returning the config
+  // of the first spout or bolt found with the name.
+  for (sp_int32 i = 0; i < _topology.spouts_size(); ++i) {
+    if (_topology.spouts(i).comp().name() == _component_name) {
+      if (_topology.spouts(i).comp().has_config()) {
+        const proto::api::Config& config = _topology.spouts(i).comp().config();
+        ConvertConfigToKVMap(config, retval);
+      }
+      return;
+    }
+  }
+
+  for (sp_int32 i = 0; i < _topology.bolts_size(); ++i) {
+    if (_topology.bolts(i).comp().name() == _component_name) {
+      if (_topology.bolts(i).comp().has_config()) {
+        const proto::api::Config& config = _topology.bolts(i).comp().config();
+        ConvertConfigToKVMap(config, retval);
+      }
+      return;
+    }
+  }
+}
+
+// Update component level config
+void TopologyConfigHelper::SetComponentConfig(proto::api::Topology* _topology,
+                                              const sp_string& _component_name,
+                                              const std::map<sp_string, sp_string>& _update) {
+  // We are assuming component names are unique and updating config for all instances
+  // with the specific component name.
+  for (sp_int32 i = 0; i < _topology->spouts_size(); ++i) {
+    proto::api::Component* comp = _topology->mutable_spouts(i)->mutable_comp();
+    if (comp->name() == _component_name && comp->has_config()) {
```

Comment:
why the check comp->has_config?

Reply:
so that in case user has a typo, topology can reject the request.

Reply:
If I read the code correctly, if a particular component doest not have a comp specific configuration initially, I wont be able to add something specific there. Why this limitation?

Reply:
Yeah. That is correct. This check is just for protection. I am open for other thoughts.

Here are a few cases,
- the topology has a config topology.config.1, and user can use command line tool to update the value. in case there is a typo like topolog.config.1, we can reject the config change.
- This could also be useful when two configs work together and need to be updated at the same time. If one of them has a typo, the update command should be cancelled.
- Or if the component name is wrong, the issue can be detected instead of waiting for the topology to react.

Like you said, the limit is that the config has to exist. I feel it is safer this way because if you want to keep the runtime config in future, you have to update the config manually in code. Requiring the config to exist could avoid some mistakes.

The check doesn't cover 100% cases and the limit is kinda artificial. If you guys prefer not having the check or any other idea, I am totally happy to consider.

Reply:
On the other hand, we can skip the check and leave all runtime configs pass through. In addition to user mistakes, there might be some unused configs and it could be harder to track them.